producer/config/writer.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"time"
+
+	"github.com/m3db/m3cluster/client"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3msg/producer/writer"
+	"github.com/m3db/m3msg/protocol/proto"
+	"github.com/m3db/m3msg/topic"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/pool"
+	"github.com/m3db/m3x/retry"
+)
+
+// ConnectionConfiguration configs the connection options.
+type ConnectionConfiguration struct {
+	DialTimeout     *time.Duration       `yaml:"dialTimeout"`
+	KeepAlivePeriod *time.Duration       `yaml:"keepAlivePeriod"`
+	ResetDelay      *time.Duration       `yaml:"resetDelay"`
+	Retry           *retry.Configuration `yaml:"retry"`
+	WriteBufferSize *int                 `yaml:"writeBufferSize"`
+	ReadBufferSize  *int                 `yaml:"readBufferSize"`
+}
+
+// NewOptions creates connection options.
+func (c *ConnectionConfiguration) NewOptions(iOpts instrument.Options) writer.ConnectionOptions {
+	opts := writer.NewConnectionOptions()
+	if c.DialTimeout != nil {
+		opts = opts.SetDialTimeout(*c.DialTimeout)
+	}
+	if c.KeepAlivePeriod != nil {
+		opts = opts.SetKeepAlivePeriod(*c.KeepAlivePeriod)
+	}
+	if c.ResetDelay != nil {
+		opts = opts.SetResetDelay(*c.ResetDelay)
+	}
+	if c.Retry != nil {
+		opts = opts.SetRetryOptions(c.Retry.NewOptions(iOpts.MetricsScope()))
+	}
+	if c.WriteBufferSize != nil {
+		opts = opts.SetWriteBufferSize(*c.WriteBufferSize)
+	}
+	if c.ReadBufferSize != nil {
+		opts = opts.SetReadBufferSize(*c.ReadBufferSize)
+	}
+	return opts.SetInstrumentOptions(iOpts)
```

Comment:
Set the instrument options upfront?

Reply:
I was trying to keep the same order as the options defined in the interface, can change if you feel strongly

Reply:
Fine either way, but it might help to prevent the issue seen earlier if the instrument options were set upfront. Your call.